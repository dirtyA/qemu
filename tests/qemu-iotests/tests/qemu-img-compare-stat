#!/usr/bin/env python3
#
# Test qemu-img compare --stat
#
# Copyright (c) 2021 Virtuozzo International GmbH.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import iotests
from iotests import qemu_img_create, qemu_io, qemu_img_log, log

iotests.script_initialize(supported_fmts=['qcow2'])

a, b, c = iotests.file_path('a', 'b', 'c')

log('= compare two images =\n')

qemu_img_create('-f', iotests.imgfmt, a, '1M')
qemu_img_create('-f', iotests.imgfmt, b, '1M')

# equal data and zero
qemu_io('-c', 'write -z 0 64K', a)
qemu_io('-c', 'write -P 0 0 64K', b)

# different data
qemu_io('-c', 'write -P 1 64K 64K', a)
qemu_io('-c', 'write -P 0 64K 64K', b)

# equal data
qemu_io('-c', 'write -P 2 128K 64K', a)
qemu_io('-c', 'write -P 2 128K 64K', b)

# equal unallocated and allocated zero
qemu_io('-c', 'write -z 192K 64K', b)

# unequal data and unallocated zero
qemu_io('-c', 'write -P 3 256K 64K', a)

qemu_img_log('compare', '--stat', a, b)

log('\n= compare two increments =\n')

qemu_img_create('-f', iotests.imgfmt, a, '1M')
qemu_img_create('-f', iotests.imgfmt, '-b', a, '-F', iotests.imgfmt, b, '1M')
qemu_img_create('-f', iotests.imgfmt, '-b', b, '-F', iotests.imgfmt, c, '1M')

qemu_io('-c', 'write -P 1 0 1M', a)
qemu_io('-c', 'write -P 2 0 64K', b)
qemu_io('-c', 'write -P 3 64K 64K', c)
qemu_img_log('compare', '--stat', b, c)

log('\n= compare two increments with --shallow=\n')
qemu_img_log('compare', '--stat', '--shallow', b, c)

log('\n= compare images of different size =\n')
qemu_img_create('-f', iotests.imgfmt, a, '1M')
qemu_img_create('-f', iotests.imgfmt, b, '2M')
qemu_io('-c', 'write -P 1 0 1M', a)
qemu_io('-c', 'write -P 2 0 1M', b)
qemu_io('-c', 'write -P 1 1M 64K', b)
qemu_io('-c', f'write -z {1024 + 64 * 2}K 64K', b)
qemu_io('-c', f'write -P 0 {1024 + 64 * 3}K 64K', b)
qemu_img_log('compare', '--stat', a, b)

log('\n= compare images with only 512 bytes different =\n')
qemu_img_create('-f', iotests.imgfmt, a, '1M')
qemu_img_create('-f', iotests.imgfmt, b, '1M')
qemu_io('-c', 'write -P 1 0 1M', a)
qemu_io('-c', 'write -P 2 0 512', b)
qemu_io('-c', f'write -P 1 512 {1024 * 1024 - 512}', b)
qemu_img_log('compare', '--stat', a, b)

log('\n= compare images with only 512 bytes different, block-size=4K =\n')
qemu_img_log('compare', '--stat', '--block-size', '4K', a, b)

log('\n= end =')
