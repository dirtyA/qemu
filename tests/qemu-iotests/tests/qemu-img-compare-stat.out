= compare two images =

Compare stats:
Key
D: DATA
Z: ZERO
A: ALLOCATED
E: after end of file

Equal:
_Z__ -> _Z__ 720896 bytes (704 KiB) 68.8%
_Z__ -> _ZA_ 65536 bytes (64 KiB) 6.2%
D_A_ -> D_A_ 65536 bytes (64 KiB) 6.2%
_ZA_ -> D_A_ 65536 bytes (64 KiB) 6.2%

Unequal:
D_A_ -> _Z__ 65536 bytes (64 KiB) 6.2%
D_A_ -> D_A_ 65536 bytes (64 KiB) 6.2%


= compare two increments =

Compare stats:
Key
D: DATA
Z: ZERO
A: ALLOCATED
E: after end of file

Equal:
D_A_ -> D_A_ 983040 bytes (960 KiB) 93.8%

Unequal:
D_A_ -> D_A_ 65536 bytes (64 KiB) 6.2%


= compare two increments with --shallow=

Compare stats:
Key
D: DATA
Z: ZERO
A: ALLOCATED
E: after end of file

Equal:
_Z__ -> _Z__ 917504 bytes (896 KiB) 87.5%

Unequal:
_Z__ -> D_A_ 65536 bytes (64 KiB) 6.2%
D_A_ -> _Z__ 65536 bytes (64 KiB) 6.2%


= compare images of different size =

Warning: Image size mismatch!
Compare stats:
Key
D: DATA
Z: ZERO
A: ALLOCATED
E: after end of file

Equal:
_Z_E -> _Z__ 851968 bytes (832 KiB) 40.6%
_Z_E -> D_A_ 65536 bytes (64 KiB) 3.1%
_Z_E -> _ZA_ 65536 bytes (64 KiB) 3.1%

Unequal:
D_A_ -> D_A_ 1048576 bytes (1 MiB) 50.0%
_Z_E -> D_A_ 65536 bytes (64 KiB) 3.1%


= compare images with only 512 bytes different =

Compare stats:
Key
D: DATA
Z: ZERO
A: ALLOCATED
E: after end of file

Equal:
D_A_ -> D_A_ 983040 bytes (960 KiB) 93.8%

Unequal:
D_A_ -> D_A_ 65536 bytes (64 KiB) 6.2%


= compare images with only 512 bytes different, block-size=4K =

Compare stats:
Key
D: DATA
Z: ZERO
A: ALLOCATED
E: after end of file

Equal:
D_A_ -> D_A_ 1044480 bytes (0.996 MiB) 99.6%

Unequal:
D_A_ -> D_A_ 4096 bytes (4 KiB) 0.4%


= end =
